rkyv/src/with.rs

```diff
@@ -473,6 +473,57 @@ pub struct AsVec;
 #[derive(Debug)]
 pub struct Niche;
 
+/// A wrapper that niches based on a generic [`Decider`].
+///
+/// A common type combination is `Option<Box<T>>`. By using a null pointer, the
+/// archived version can save some space on-disk.
+///
+/// # Example
+///
+/// ```
+/// use core::mem::size_of;
+///
+/// use rkyv::{
+///     niche::decider::{NaN, Null},
+///     with::Nicher,
+///     Archive, Archived,
+/// };
+///
+/// #[derive(Archive)]
+/// struct BasicExample {
+///     maybe_box: Option<Box<str>>,
+///     maybe_non_nan: Option<f32>,
+/// }
+///
+/// #[derive(Archive)]
+/// struct NichedExample {
+///     #[rkyv(with = Nicher<Null>)]
+///     maybe_box: Option<Box<str>>,
+///     #[rkyv(with = Nicher<NaN>)]
+///     maybe_non_nan: Option<f32>,
+/// }
+///
+/// assert!(
+///     size_of::<Archived<BasicExample>>()
+///         > size_of::<Archived<NichedExample>>()
+/// );
+/// ```
+///
+/// [`Decider`]: crate::niche::decider::Decider
+pub struct Nicher<D: ?Sized>(PhantomData<D>);
```

Comment:
I did some tests and, as far as I can tell, we could add a decider type parameter to the existing `Niche` wrapper _as long as_ we provide a default. It seems reasonable to add a `Default` decider which only implements `Decider` for "reasonable" types like boxes and the non-zero types. This would let us keep the ergonomics of `#[rkyv(with = Niche)]` while still allowing the flexibility of `#[rkyv(with = Niche<One>)]`.

Reply:
I was also thinking of how to combine `Niche` with the new generic version but wanted approval for this whole approach first 😄 

It would definitely be possible to add a `DefaultNiching` to replace the two nichings that were already supported (`Box<T>` and `NonZero*`) and then turn the `Niche` struct into a type alias for `Nicher<DefaultNiching>`. The small downside is that `Niche` would no longer be a unit struct so that's technically a breaking change. The bigger downside is that the niched archived types would be replaced with the new `NichedOption` type which is most definitely a breaking change.

I think the better solution would be to mark `Niche` as deprecated with an info about `Nicher` and then remove it entirely for the next breaking release.

Reply:
Ah, that's a great point. I agree with your suggestion, let's land this without changing anything about `Niche`. I'll deprecate it before the next point release.